toolchain/sem_ir/inst_kind.h

```diff
@@ -35,28 +32,47 @@ enum class InstValueKind : int8_t {
+// to define a constant value.
+//
+// This specifies whether an instruction of this kind can have a corresponding
+// constant value in the `constant_values()` list, and whether an instruction of
+// this kind can be added to the `constants()` list.
 enum class InstConstantKind : int8_t {
-  // This instruction never defines a constant value. For example,
-  // `UnaryOperatorNot` never defines a constant value; if its operand is a
-  // concrete constant, its constant value will instead be a `BoolLiteral`. This
-  // is also used for instructions that don't produce a value at all.
+  // This instruction is never constant. Its constant value is always
+  // `NotConstant`. This is also used for instructions that don't produce a
+  // value at all and aren't used as constants.
   Never,
-  // This instruction may be a symbolic constant, depending on its operands, but
-  // is never a concrete constant. For example, a `Call` instruction can be a
-  // symbolic constant but never a concrete constant.
+  // This instruction never defines a constant value, but can evaluate to a
+  // constant value of a different kind. For example, `UnaryOperatorNot` never
+  // defines a constant value; if its operand is a concrete constant, its
+  // constant value will instead be a `BoolLiteral`. This is the default.
+  Indirect,
+  // This instruction may define a symbolic constant, depending on its operands,
+  // but never a concrete constant. For example, a `Call` instruction can define
+  // a symbolic constant but never a concrete constant. The instruction may have
+  // a concrete constant value of a different kind.
   SymbolicOnly,
   // This instruction can define a symbolic or concrete constant, but might not
-  // have a constant value, depending on its operands. For example, a
-  // `TupleValue` can define a constant if its operands are constants.
+  // have a constant value, might have a constant value that is not defined by
+  // itself, or might result in a compile-time error, depending on its operands.
+  // For example, `ArrayType` is a compile-time constant if its operands are
+  // constant and its array bound is within a valid range.
   Conditional,
-  // This instruction always has a constant value of the same kind. For example,
-  // `IntValue`.
+  // This instruction defines a symbolic or concrete constant whenever its
+  // operands are constant. Otherwise, it is non-constant. For example, a
+  // `TupleValue` defines a constant if and only if its operands are constants.
+  // Constant evaluation support for types with this constant kind is provided
+  // automatically.
+  WheneverPossible,
+  // This instruction always has a constant value of the same kind. This is the
+  // same as `WheneverPossible`, except that the operands are known in advance
+  // to always be constant. For example, `IntValue`.
   Always,
+  // This instruction is a unique constant. This is used for declarations whose
```

Comment:
This instruction *produces* a unique constant? It seems to be used for declarations, where each instance has different parameters and produces a distinct value that is itself. It does not seem to be used for the even more unique case of singleton instructions, which use `Always` instead. Could you explain in your comment the differences between these two enum values?

Reply:
I mean this instruction itself *is* the definition of a unique constant. I've extended the comment to try to clarify.